### PR TITLE
ure.sub

### PR DESCRIPTION
**Pull request by dmazzella**
Added sub into ure module and test

unix/main.c  and extmod/modmachine.c modified for build under MSYS2

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the effort, but your code does not work correctly.  Even on the simplest cases it fails, eg:
```
print(re.sub('a', 'A', 'aa'))
print(re.sub('a', 'A', 'ab'))
```
CPython prints `AA` and `Ab`, whilst your patch prints `A` and `A`.  The problem is that you are only calling re_exec once, when you should call it many times over the input string.

File: unix/main.c [hunk 1]
```diff
@@ -416,7 +416,7 @@ int main(int argc, char **argv) {
             } else if (strcmp(argv[a], "-v") == 0) {
                 mp_verbose_flag++;
             } else if (strncmp(argv[a], "-O", 2) == 0) {
-                if (isdigit(argv[a][2])) {
+                if (isdigit((byte)argv[a][2])) {
```
Comment (dpgeorge) [inline unix/main.c]: What error/warning is given by MSYS2 without this fix?

Comment (dmazzella) [inline unix/main.c]: Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
CC main.c
main.c: In function ‘main’:
main.c:419:17: error: array subscript has type ‘char’ [-Werror=char-subscripts]
                 if (isdigit(argv[a][2])) {
                 ^
cc1: all warnings being treated as errors
../py/mkrules.mk:47: recipe for target 'build/main.o' failed
make: *** [build/main.o] Error 1

Comment (dmazzella) [inline unix/main.c]: $gcc --version
gcc (GCC) 4.9.2

Comment (stinos) [inline unix/main.c]: That is strange. The error is about the array subscript, so that would be ```a``` or ```2``` on that line, being of type char. But both are int? And the error goes away by making what seems like an unrelated change? (though it is indeed the case that isdigit should not be passed signed char)

Comment (pfalcon) [inline unix/main.c]: Well, from that "fix", the issue is quite obvious - the common way to implement isdigit() is by subscripting a table of flags, which of course kabooms if char is "negative". But then it's quite a buggy isdigit(), which requires such "fix" applied to each its invocation, and it's isdigit() which requires fixing.

But deeper problem with this PR is that it packs unrelated changes, and number of commits in it is ~30, so it's hard to reason what's done for what.

Comment (stinos) [inline unix/main.c]: You should file a bug report to the standard comittee for defining isdigit as buggy :P

Comment (pfalcon) [inline unix/main.c]: No, author of this patch should bring this issue to MSYS2 or wherever he found such isdigit implementation.

Comment (stinos) [inline unix/main.c]: Well if the standard says 'undefined behaviour' the implementation is free to do whatever it wants, including kabooming all over the place. So in theory an implementation indexing a table without checking arguments is perfectly fine. Though I agree that in practice however it could just check the argument, and most definitely it cannot raise bogus implementation-defined warnings.

Comment (dpgeorge) [inline unix/main.c]: isdigit is supposed to take an int as argument, so it should be able to cope with a char without error or warning.

**Comment by dmazzella** [comment]
Hi Damien,
i have fixed the issue, now the output for CPython and MicroPython is the same for calls:

print(re.sub('a', 'A', 'aa'))
print(re.sub('a', 'A', 'ab'))

**Comment by stinos** [comment]
You should split this PR into seperate PRs (one with just the changes for ure.sub, one for 'fix compile problems for gcc version 4.9.2 (GCC) Target: x86_64-pc-msys' and one for 'Fix unused parameter 'n_args'') and rebase your changes onto the latest master instead of merging master into your branches. Then each PR will only have one, or a couple of, commits which makes it much easier to review.

**Comment by dmazzella** [comment]
OK, I close this and open a new one.
